### PR TITLE
feat: add project endpoints

### PR DESCRIPTION
**Pull request by jan-zabloudil**
This pull request introduces the projects. The functionality for 'project invitations' and 'project members' will be added in separate pr.

### CODE REVIEW

File: service/model/project.go
Comment (jan-zabloudil) [inline service/model/project.go:20]: Project contains `ReleaseAlertConfig` and `SlackConfig`.

`ReleaseAlertConfig` replaces the original concept of release message templates and determinates which information to include into release alert. I also decided to use `ReleaseAlert` instead of `ReleaseMessage`.

`SlackConfig` has only one field for now - channel id.

File: service/model/release_alert.go
Comment (jan-zabloudil) [inline service/model/release_alert.go:1]: For now only `ReleaseAlertConfig` is only used as part of Project model.

But I decided to define `ReleaseAlert` already in order to better imagine how ReleaseAlertConfig will be used. And therefore I created separete file `/service/model/release_alert`.

`ReleaseAlert` holds the alert data. `ReleaseMessageBuilder` will be used for building the alert and will accept `ReleaseAlertConfig` as param.

File: service/project.go
Comment (jan-zabloudil) [inline service/project.go:1]: Project service functions do not have any authorization logic yet because project members feature will be implemented in the separate PR.

File: service/model/project_test.go
Comment (jan-zabloudil) [inline service/model/project_test.go:1]: I only added tests for service model functions.

Do you think it also make sense to mock repositories and write tests for service functions?

File: service/model/project.go
Comment (jan-zabloudil) [inline service/model/project.go:81]: Validate function enforces all business rules and is called when creating or updating project object. 

I prefer to not use any validator package in service layer.

File: repository/model/project.go
Comment (jan-zabloudil) [inline repository/model/project.go:77]: When converting db object into service object I also perform validation

File: transport/model/project.go
Comment (jan-zabloudil) [inline transport/model/project.go:1]: I prefer to have a separate struct for each operation, such as CreateProjectRequest, UpdateProjectRequest, and ProjectResponse.

In the transport layer, I only perform input validation to ensure that the user submits the correct payload. Any business rules are validated in the service layer

File: service/model/release_alert.go
Comment (jan-zabloudil) [inline service/model/release_alert.go:1]: `ReleaseAlertConfig` is a part of the Project struct, currently not used outside this context. 

But to illustrate the potential use of Config, I already have created a `ReleaseAlert` struct and a builder for it. And therefore I put all the related code into `/service/model/release_alert`. The ReleaseAlert struct is designed to store alert data. The builder, which takes a Config as a parameter, builds the alert.

File: service/project.go
Comment (jan-zabloudil) [inline service/project.go:1]: There is no authorization logic in service functions for now because project members feature will be added in separate PR.

File: service/model/project_test.go
Comment (jan-zabloudil) [inline service/model/project_test.go:1]: I created tests only for service model functions for now.

Would you also mock repositories and write tests for service functions?

File: transport/utils/validate.go
Comment (bolshakoff) [inline transport/utils/validate.go:25]: A bit confused about this change; especially because the `init()` function is not used anywhere.. 🤔

File: supabase/migrations/20240328123715_projects.sql
Comment (jan-zabloudil) [reply supabase/migrations/20240328123715_projects.sql]: I understand your point, but the issue is that Supabase does not support transactions, which complicates scenarios where you want to insert data into multiple tables. 
To achieve this, I would need to write a PostgreSQL function, which is not very maintainable. In this case, I feel that having the configuration saved in JSON is sufficient, but I will discuss it further with Marek during our call. Thank you!

File: supabase/migrations/20240328123715_projects.sql
Comment (jan-zabloudil) [reply supabase/migrations/20240328123715_projects.sql]: It was auto-generated migration. You are right, I fixed it

File: transport/utils/validate.go
Comment (jan-zabloudil) [reply transport/utils/validate.go:25]: Should be called automatically when i import the package.

This change ensures that when validation errors occur, the field names in the error messages correspond to the JSON tags of the fields, which is more user-friendly.

**Comment by jan-zabloudil** [comment]
This branch is now rebased onto https://github.com/jan-zabloudil/release-manager/pull/16 because during review I also rebased it to work with more recent code.

File: pkg/dberrors/dberrors.go
Comment (jan-zabloudil) [inline pkg/dberrors/dberrors.go:50]: Not related to this PR but addresses an issue left over from a previous PR.

File: service/model/user_role.go
Comment (jan-zabloudil) [inline service/model/user_role.go:12]: New role `Manager` introduces. Why?

- Admin  role - have access to organization settings (implemented in #14) and can create projects as well

- Manager role - does not have access to organization settings, but can create projects _(role for Engineering managers in STRV)_

Other naming option would be `SuperAdmin` and `Admin roles`.

If you approve the addition of this new role, I will also create an endpoint that enables admin users to assign the manager role to other users.

File: service/project.go
Comment (jan-zabloudil) [inline service/project.go:39]: Authorization will be added in the next PR, where project members feature will be added.

File: service/model/project.go
Comment (jan-zabloudil) [inline service/model/project.go:45]: `ReleaseAlertConfig` defines the information that will be included in a release alerts for a given project. 

I chose 'ReleaseAlert' over 'ReleaseMessage' as I believe it’s a more suitable name

File: service/model/user_role.go
Comment (bolshakoff) [reply service/model/user_role.go:12]: Interesting point! I'd definitely steer towards something more simple and obvious, for example:
- org admin
- project admin
- user
How about other similar platforms? What names are they using?

And do we expect to have more different items?

If yes, consider simply having users and rights (read/modify org/project).

File: service/model/project.go
Comment (bolshakoff) [reply service/model/project.go:45]: `Alert` sounds too negative. :)
Also agree that `message` is a bit too neutral and general.
How about `notification`?

Comment (jan-zabloudil) [reply service/model/project.go:45]: Good point, let's go with `ReleaseNotification`

**Comment by jan-zabloudil** [comment]
When discussing the release management app proposal, we considered two approaches for organizing releases:

**Project > Releases**
- originally proposed by @CermakM.
- each project represents a single application (e.g., Backend app for project X).
- one project = one GitHub repository.
- I suggest renaming it to `App > Releases`.
 
**Project > Apps > Releases**

I initially favored the second approach (Project > Apps > Releases) and built the entire proposal around this structure. But I have realized that it introduces unnecessary complexity without offering significant benefits. The first suggestion is much more flexible and aligns better with company workflows (including STRV).

**Why?**
Different teams typically handle different platforms for given project. For example, the Go team manages the backend solution, the iOS team handles the iOS app, and the Android team takes care of the Android app.

It makes sense for each Engineering Manager to create an app within ReleaseManager for which they are responsible. They can then invite engineers from their team with permissions to create releases. Managers can also invite engineers from other teams or PMs but only grant them read permissions. The manager will also connect the app to a GitHub repository and configure release notifications. 

But then, do we really need the project level at all? 
It seems that adding the project level would not bring any significant benefits and would only increase complexity. Adopting this approach would require managing permissions and settings on two levels. Or keep settings/permissions only on one level but significantly decrease the release manager app’s flexibility.

Since we are developing an open-source MVP, I would definitely choose the App > Releases solution.

File: pkg/apierrors/apierrors.go
Comment (CermakM) [inline pkg/apierrors/apierrors.go:78]: Just a suggestion here, but as you can see, the `IsError` functions have all the same definition and the only thing they differ in are the error codes comparisons. Meaning you can simplify this flow by implementing one `IsErrorWithCode` function of the signiture `IsErrorWithCode(err error, code Code)`.

File: pkg/responseerrors/responseerrors.go
Comment (CermakM) [inline pkg/responseerrors/responseerrors.go:12]: I would rather structure the error codes as `errCodeX` rather than `xErrorCode`, this makes it easier to grep them both visually and via autocomplete.

File: repository/model/project.go
Comment (CermakM) [inline repository/model/project.go:17]: Why not `time.Time`?

File: repository/project.go
Comment (CermakM) [inline repository/project.go:27]: Is there any point in returning the same structure as passed as a parameter? To me it looks like this is an unnecessary overhead and simply returning an `error` would be enough.

File: service/model/project.go
Comment (CermakM) [inline service/model/project.go:71]: Define a variable `now` and use it for both fields, otherwise they might not be equal, which is weird considering no update happened.

**Review by CermakM — CHANGES_REQUESTED** [review]

**Comment by CermakM** [comment]
Thanks for the elaboration on the entity structure! I 💯 agree 👍 

Regarding the user permissions:

> AppAdmin / ProjectAdmin
>
> Can create apps
  Does not have access to organization settings
  This role would typically be assigned to Engineering Managers
  
In the current STRV setting, I believe it's okay for us to have only organization admin and project member structure. However, from a long term perspective, I would absolutely consider having a project admin. Especially if the organization-wide settings grows in complexity and might potentially hold sensitive info. Also, since project administrators may not be technically savvy, there is some risk in allowing them access to org-wide settings.

File: repository/model/project.go
Comment (jan-zabloudil) [reply repository/model/project.go]: Thanks! Initially, I had problems with marshaling/unmarshaling into a time.Time variable. But that was probably due to some beginner's mistakes. It works fine with `time.Time` for `timestamp with timezone` column 👍

**Comment by jan-zabloudil** [comment]
Thank you 👍 I'm reassigning the PR to myself and will implement the remaining functionalities for `apps`.

**Comment by jan-zabloudil** [comment]
- Environments have been added.

To keep PRs smaller, the following features will be implemented in separate PRs:
- Project members
- Connection to GitHub repository (it would be good to merge PR #14  first to have access to the GitHub token).

In the end, I also decided to continue referring to it as `Project` instead of `App`. It is more general, but I think it’s not necessarily a bad thing. Moreover, `project members` works better than `app members`, etc.

File: pkg/url/url.go
Comment (jan-zabloudil) [inline pkg/url/url.go:10]: I'm not sure about this... I at least wanted to perform a basic check to see if a URL is valid according to a 'human' definition of a "correct" URL (of course, many incorrect ones would still pass). 

Or it might be better to use some existing validator...

**Comment by jan-zabloudil** [comment]
Sorry, I've realized I need to implement a different approach to environments. Reassigning it to myself again and I'm on it.

File: pkg/url/url.go
Comment (bolshakoff) [reply pkg/url/url.go:10]: Yes, I think it can be done using the `validator` package.

**Comment by jan-zabloudil** [comment]
> Sorry, I've realized I need to implement a different approach to environments. Reassigning it to myself again and I'm on it.

In STRV, all projects have development, staging, and production environments. However, I have decided not to have fixed environments to maintain flexibility, especially since I am developing an open source app.

1. When a manager creates a project, they can set up environments; the name field is required, while the URL and release fields are optional. _Potentially in STRV, we could automatically create development, staging, and production environments when a project is created_.

 2. When using the app, you can visit the environments page to see which releases are deployed to each environment, and you can quickly make changes.

Projects _ Environments

You can also "deploy" a release to an environment from the release page.

Projects _ Releases _ Add

File: pkg/url/url.go
Comment (jan-zabloudil) [reply pkg/url/url.go:10]: Thanks, I am going to change it in separate PR once this one is merged 👍

File: service/model/environment.go
Comment (CermakM) [inline service/model/environment.go:51]: Why do I need to provide this URL? It should be accesible on the base url + id, shouldn't it?

File: service/environment.go
Comment (CermakM) [inline service/environment.go:30]: Shouldn't this be only the admin's privilage?

File: service/environment.go
Comment (CermakM) [inline service/environment.go:110]: Same here, project admin only I would say.

File: supabase/migrations/20240417150340_environments.sql
Comment (CermakM) [inline supabase/migrations/20240417150340_environments.sql:4]: Should be unique per project.

File: service/environment.go
Comment (CermakM) [inline service/environment.go:19]: I think it's an unnecessary overhead to have environment creation as a separate service. It's part of the project resource, doesn't have a seaparate resource endpoint, hence I would put it under the project service. Not that you don't have to have a service per resource either, rather structure it per domain.

**Review by CermakM — CHANGES_REQUESTED** [review]

File: pkg/dberrors/dberrors.go
Comment (jan-zabloudil) [reply pkg/dberrors/dberrors.go]: In the end I dont need that error anymore, I removed it.

File: service/model/environment.go
Comment (jan-zabloudil) [reply service/model/environment.go:42]: Renamed to `ServiceURL` to improve clarity.

File: service/project.go
Comment (bolshakoff) [reply service/project.go:124]: Sorry, consider this resolved..

But actually..
I'd return the duplicate name error here; ideally simplify it to return `(*bool, error)`.
A bit more code, but more obvious, more readable.

Comment (bolshakoff) [reply service/project.go:124]: I mean `s.validateEnvironmentNameUnique()`

Comment (jan-zabloudil) [reply service/project.go:115]: Thank you for review, all very good points!

Just here I chose to return `(bool, error)`.
I understand that by returning `(*bool, error)`, you can distinguish the following cases based on bool value:
- true - the name is unique
- false - the name is not unique
- nil - it cannot be determined if the name is unique

But in case of `(bool, error)` I simply check if err is nil. If it isn’t, I know that the uniqueness couldn't be determined due to an error. If err is nil, I can use the bool value directly without needing to dereference it.

File: service/model/environment.go
Comment (jan-zabloudil) [reply service/model/environment.go]: > (optional) And actually, it might be even more clean to have the following signature:
> 
>     * `Update(u EnvironmentUpdate) (*Environment, error)`

Not sure if I get the point. What is the benefit of returning a pointer to the Environment in the case of a pointer receiver? Is it just for better readability?

```
env, err := env.Update(...)
```

**Review by CermakM — APPROVED** [review]

**Pull request merged**
